### ISSUE TITLE
Failure to link with device code using assert

### ISSUE THREAD

**Issue by markdewing** [llvm/llvm-project#57656]
Code that uses `assert` in device regions fails to compile at the link step with the error:
```console
nvlink error   : Undefined reference to '__assert_fail' in '/tmp/test_assert-da8852-nvptx64-nvidia-cuda-sm_61-108a4e.cubin'
```

The test code is
```cpp
#include <cassert>

int main()
{
    int i = 1;
#pragma omp target
    assert(i > 0);
    return 0;
}
```
Compiled with
```shell
clang++  -O2 -fopenmp -fopenmp-targets=nvptx64-nvidia-cuda  test_assert.cpp
```

This issue affects LLVM 15.  It compiles without complaint on LLVM 14 (and the assert triggers as expected if the value of `i` is changed to -1)
The same behavior happens with an AMD target (works with 14, fails to link on 15)

**Comment by jdoerfert** [comment]
@jhuber6 Can you take a look, plz.

**Comment by jhuber6** [comment]
> @jhuber6 Can you take a look, plz.

It doesn't happen when using LTO, my guess is we're dropping the declaration somewhere when we link per-TU.

**Comment by jhuber6** [comment]
Created a fix in [D133594](https://reviews.llvm.org/D133594), let me know if it solves your problem.
